de.fhdo.lemma.servicedsl.extractor/src/de/fhdo/lemma/servicedsl/extractor/ServiceDslExtractor.xtend

```
@@ -0,0 +1,202 @@
+            case SYNCHRONOUS: "sync"
+            default: null
+        }
+    }
+
+    private def generate(ExchangePattern pattern) {
+        return switch(pattern) {
+            case IN: "in"
+            case OUT: "out"
+            case INOUT: "inout"
+            default: null
+        }
+    }
+
+    private def generate(ImportedServiceAspect aspect) {
+        return '''@«aspect.importedAspect?.technology?.name»::«
+        FOR s : aspect.importedAspect.getQualifiedNameParts(false, true) SEPARATOR '.'»«s
+        »«ENDFOR»'''
+    }
+
+    private def generateType(Parameter parameter) {
+        return if (parameter.primitiveType !== null)
+                (parameter.primitiveType as PrimitiveType).generate
+            else
+                parameter.importedType.generate
+    }
+
+    private def generate(PrimitiveType type) {
+        return type.typeName
+    }
+    private def generateTechAnnotation(Microservice service) {
+        return '''«FOR tech :
+            service.serviceModel.imports.filter[it.importType == ImportType.TECHNOLOGY]
+            »@technology(«tech.name»)«ENDFOR»'''
+    }
+    private def generate(Protocol protocol) {
+        val techName = protocol.technology?.name ?: '''[PROTOCOL TECHNOLOGY URI NOT DEFINED]'''
+        return '''«techName»::«FOR p : protocol.qualifiedNameParts SEPARATOR '.'»«p»«ENDFOR»'''
+    }
+
+    private def generate(ImportedType importedType) {
+        switch (importedType.import.importType) {
+            case ImportType.TECHNOLOGY: {
+                //TODO is this ever triggered?
+                return '''«importedType.import.name»::«importedType.type»'''
+            }
+            case ImportType.DATATYPES: {
+                val importedTypeName = (importedType.type as ComplexType).buildQualifiedName(".")
+                return '''«importedType.import.name»::«importedTypeName»'''
+            }
+            default: {
+                return '''[COULD NOT MATCH IMPORTED TYPE]'''
```

Would be better to throw an Exception for this case since it is `impossible to occur(tm)`. An Exception would therefore clearly indicate that something went very wrong.